### PR TITLE
Redo parsing of look-ahead vectors

### PR DESCRIPTION
**Pull request by rsmmr**
No test for the problem we used to have as I'm not sure how to test for that reliably (we'd need to trigger stack overflow)

### CODE REVIEW

File: spicy/toolchain/include/compiler/detail/codegen/productions/while.h [hunk 1]
```diff
@@ -8,33 +8,70 @@
...
 #include <spicy/compiler/detail/codegen/production.h>
+#include <spicy/compiler/detail/codegen/productions/look-ahead.h>
 
 namespace spicy::detail::codegen::production {
 
 /**
- * A production executing as long as a given boolean expression evaluates
- * to true.
+ * A production executing as long as either a given boolean expression evaluates
+ * to true, or, if no expression is provided, as determined by look-ahead symbols
```
Comment (bbannier) [inline spicy/toolchain/include/compiler/detail/codegen/productions/while.h:17]:
```suggestion
 * to true, or, if no expression is provided, as determined by look-ahead symbols.
```

File: spicy/toolchain/include/compiler/detail/codegen/productions/while.h [hunk 2]
```diff
@@ -8,33 +8,70 @@
...
     bool eodOk() const { return nullable(); }
     bool atomic() const { return false; }
     bool supportsSynchronize() const { return hasSize() || _body.supportsSynchronize(); }
-    std::string render() const { return hilti::util::fmt("while(%s): %s", _expression, _body.symbol()); }
+    std::string render() const;
 
 private:
-    Expression _expression;
     Production _body;
+    std::optional<Expression> _expression;
```
Comment (bbannier) [inline spicy/toolchain/include/compiler/detail/codegen/productions/while.h:73]: Let's add a comment here that only one of these should be set.

File: spicy/toolchain/include/compiler/detail/codegen/productions/while.h [hunk 3]
```diff
@@ -8,33 +8,70 @@
...
+     */
     While(const std::string& symbol, Expression e, Production body, const Location& l = location::None)
-        : ProductionBase(symbol, l), _expression(std::move(e)), _body(std::move(body)) {}
+        : ProductionBase(symbol, l), _body(std::move(body)), _expression(std::move(e)) {}
 
-    const Expression& expression() const { return _expression; }
-    const Production& body() const { return _body; }
+    /**
+     * Constructor for a while-loop using look-ahead as the condition for
+     * termination. When using this constructor, `preprocessLookAhead()` must
```
Comment (bbannier) [inline spicy/toolchain/include/compiler/detail/codegen/productions/while.h:29]: Could we instead pass the grammar into this constructor to make it easier to guarantee that exactly on of `_expression` or `_body_for_grammar` is set?

File: spicy/toolchain/include/compiler/detail/codegen/productions/while.h (hunk 2 above)
Comment (rsmmr) [reply spicy/toolchain/include/compiler/detail/codegen/productions/while.h:73]: Actually `_body` is always set.

File: spicy/toolchain/include/compiler/detail/codegen/productions/while.h (hunk 3 above)
Comment (rsmmr) [reply spicy/toolchain/include/compiler/detail/codegen/productions/while.h:29]: AST classes should be instantiable without having a grammar (yet), that's why this is a little bit awkward.

File: spicy/toolchain/include/compiler/detail/codegen/productions/while.h (hunk 2 above)
Comment (bbannier) [reply spicy/toolchain/include/compiler/detail/codegen/productions/while.h:73]: I meant that only one of `_expression` or `_body_for_grammar` should be set making this some sort of union type.

**Pull request merged**